### ISSUE TITLE
upload data to product model

### ISSUE THREAD

**Issue by kdionisio** [HumanExposure/factotum#467]
As a Factotum data curator
I would like the ability to batch upload product data for products that do not yet exist
So that I can efficiently load extracted data specific to products

Definition of Done:
1. Ability on datagroup page to batch upload product data
  - product data upload csv should accept the following columns:
    - data document ID (required; this is only column that is required to be populated)
     - dashboard_product.upc
     - dashboard_product.url
     - dashboard_product.brand_name
     - dashboard_product.size
     - dashboard_product.color
     - dashboard_product.item_id
     - dashboard_product.parent_item_id
     - dashboard_product.short_description
     - dashboard_product.long_description
     - dashboard_product.thumb_image
     - dashboard_product.medium_image
     - dashboard_product.large_image
     - dashboard_product.model_number
     - dashboard_product.manufacturer
2. Upon upload, new product(s) are created associated with the data document indicated in data document ID within the upload
3. If no UPC is included in the upload, a stub UPC is created when the new product record is created
4. If a UPC is included in the upload, application must check to ensure that UPC does not already exists. If UPC already exists, record is rejected. Upon completion of the upload, user sees a message listing all data document IDs for which the upload failed due to already existing UPCs (ie the 'rejected records').
5. Download template for above product upload csv is available on data group page; downloadable template includes all above columns, plus a column for data document filename
6. Multiple products can be created for a single document. If user attempts to upload product data for a document which already has an associated product, a new product is created with the newly uploaded data (rather than trying to insert data into the existing product record). If a user uploads a product file which includes multiple rows for a single data document ID, each row should be assumed to be a different product for that document

**Comment by mwfrost** [comment]
@kdionisio Are there any Data Group types for which this feature should be disabled?

**Comment by mwfrost** [comment]
This new functionality only applies to Composition data groups

**Comment by mwfrost** [comment]
> Multiple products can be created for a single document. If user attempts to upload product data for a document which already has an associated product, a new product is created with the newly uploaded data (rather than trying to insert data into the existing product record). *If a user uploads a product file which includes multiple rows for a single data document ID, each row should be assumed to be a different product for that document*

@kdionisio Is it necessary to handle the case where two rows have the same data document ID _and_ identical product information? Because that second product record is going to get rejected if its UPC already exists.

**Comment by mwfrost** [comment]
@kdionisio The Product model uses a `title` field - it's the headline of every product page. I don't see that among the fields listed in this ticket. Was that an intentional omission?

**Comment by kdionisio** [comment]
@mwfrost re. the first question on different document IDs but identical product information - the only instance we need to check for is if a UPC already exists (so either if the UPC already exists in the database, OR, as you state, if within the upload there are two different document IDs with the same UPC). In the case of the former, the attempted upload entry should be rejected (ideally with a message to the user that 'UPC XX in row YY already exists in Factotum'). In the case of the latter, both entries should be rejected (ideally with a message to the user that document IDs X and Y contain identical UPC).

**Comment by kdionisio** [comment]
@mwfrost regarding product.title - thank you, that is a good catch! that field should be included in the upload as well

**Comment by kdionisio** [comment]
@mwfrost @mcovalt Attached is an example product data upload file if this helps with testing. Note that this file corresponds to a newly created data group on prod.[https://app.zenhub.com/files/99837877/f3e292ba-612f-4e62-9bc6-d8dc0e02d17e/download](https://app.zenhub.com/files/99837877/f3e292ba-612f-4e62-9bc6-d8dc0e02d17e/download)

**Comment by mwfrost** [comment]
@kdionisio This file won't work as-is with the page. The data documents are not in the SQL dump from production, and the column names don't match. Once the ticket is merged into production, users will be able to download a template pre-populated with the document ids and the canonical column headers.

Depending on how confident @mcovalt is with the code, we might decide to call my tests good enough without trying to upload this.

**Comment by kdionisio** [comment]
That's fine, I'm ok with pushing and trying it, then reticketing if needed. Unless you are unsure and would like us to create a mockup upload file for a data group that exists in your dump of the database

**Comment by mwfrost** [comment]
No need to do that. I'll build a more robust test.

**Comment by mwfrost** [comment]
@kdionisio Here are the cases where identical UPCs are being uploaded with different document IDs:

data document ID | dashboard_product.upc
-- | --
1478360 | 72213241105
1478357 | 72213241105
1478445 | 78254050638
1478224 | 78254050638
1478443 | 78372012143
1478435 | 78372012143
1478453 | 78372029165
1478433 | 78372029165

Your note above says "In the case of the latter, both entries should be rejected (ideally with a message to the user that document IDs X and Y contain identical UPC)."

The uploader does not currently check to make sure the file itself does not contain duplicate UPCs -- it only rejects UPCs that already exist. So in each of these examples, the first row would be added to the database, then the second row would be rejected. If you can confirm that that is not what you want, I will revise the code to preemptively reject any rows in the uploaded file that share a UPC with one another, whether or not that UPC is already in the database.

**Comment by kdionisio** [comment]
@mwfrost If your proposal is easy to implement, please go ahead and do so. If not, then you can leave this as is, and I'll add a new ticket to check for duplicate UPCs in the product upload file prior to upload.
